From 810a300398dc93b7becca4a7f286e8999adc09a5 Mon Sep 17 00:00:00 2001
From: Akira TAGOH <akira@tagoh.org>
Date: Fri, 8 Mar 2024 01:55:14 +0900
Subject: [PATCH] Sort out bitmap related config files

70-no-bitmaps-and-emoji.conf: reject bitmap fonts including bitmap emoji fonts.
70-no-bitmaps-except-emoji.conf: reject bitmap fonts except bitmap emoji fonts.
70-no-bitmaps.conf: same to 70-no-bitmaps-except-emoji.conf for backward compatibility
but deprecated. This will be dropped in the future.

Fixes https://gitlab.freedesktop.org/fontconfig/fontconfig/-/issues/409
---
 conf.d/70-no-bitmaps-and-emoji.conf    | 13 +++++++++++++
 conf.d/70-no-bitmaps-except-emoji.conf | 14 ++++++++++++++
 conf.d/70-no-bitmaps.conf              | 11 ++---------
 conf.d/Makefile.am                     |  2 ++
 conf.d/meson.build                     |  2 ++
 5 files changed, 33 insertions(+), 9 deletions(-)
 create mode 100644 conf.d/70-no-bitmaps-and-emoji.conf
 create mode 100644 conf.d/70-no-bitmaps-except-emoji.conf

diff --git a/conf.d/70-no-bitmaps-and-emoji.conf b/conf.d/70-no-bitmaps-and-emoji.conf
new file mode 100644
index 00000000..3e18a4cf
--- /dev/null
+++ b/conf.d/70-no-bitmaps-and-emoji.conf
@@ -0,0 +1,13 @@
+<?xml version="1.0"?>
+<!DOCTYPE fontconfig SYSTEM "urn:fontconfig:fonts.dtd">
+<fontconfig>
+  <description>Reject bitmap fonts, including bitmap emoji fonts</description>
+  <!-- Reject bitmap fonts -->
+  <selectfont>
+    <rejectfont>
+      <pattern>
+        <patelt name="outline"><bool>false</bool></patelt>
+      </pattern>
+    </rejectfont>
+  </selectfont>
+</fontconfig>
diff --git a/conf.d/70-no-bitmaps-except-emoji.conf b/conf.d/70-no-bitmaps-except-emoji.conf
new file mode 100644
index 00000000..3170cd6f
--- /dev/null
+++ b/conf.d/70-no-bitmaps-except-emoji.conf
@@ -0,0 +1,14 @@
+<?xml version="1.0"?>
+<!DOCTYPE fontconfig SYSTEM "urn:fontconfig:fonts.dtd">
+<fontconfig>
+  <description>Reject bitmap fonts except bitmap emoji fonts</description>
+  <!-- Reject bitmap fonts -->
+  <selectfont>
+    <rejectfont>
+      <pattern>
+        <patelt name="outline"><bool>false</bool></patelt>
+        <patelt name="scalable"><bool>false</bool></patelt>
+      </pattern>
+    </rejectfont>
+  </selectfont>
+</fontconfig>
diff --git a/conf.d/70-no-bitmaps.conf b/conf.d/70-no-bitmaps.conf
index 10203acb..767e4115 100644
--- a/conf.d/70-no-bitmaps.conf
+++ b/conf.d/70-no-bitmaps.conf
@@ -1,13 +1,6 @@
 <?xml version="1.0"?>
 <!DOCTYPE fontconfig SYSTEM "urn:fontconfig:fonts.dtd">
 <fontconfig>
-  <description>Reject bitmap fonts</description>
-<!-- Reject bitmap fonts -->
- <selectfont>
-  <rejectfont>
-   <pattern>
-     <patelt name="outline"><bool>false</bool></patelt>
-   </pattern>
-  </rejectfont>
- </selectfont>
+  <description>Reject bitmap fonts except bitmap emoji fonts (deprecated; use 70-no-bitmaps-except-emoji.conf)</description>
+  <include ignore_missing="yes">70-no-bitmaps-except-emoji.conf</include>
 </fontconfig>
diff --git a/conf.d/Makefile.am b/conf.d/Makefile.am
index fd858348..737b4fb2 100644
--- a/conf.d/Makefile.am
+++ b/conf.d/Makefile.am
@@ -98,6 +98,8 @@ template_DATA =				\
 	65-nonlatin.conf		\
 	69-unifont.conf			\
 	70-no-bitmaps.conf		\
+	70-no-bitmaps-and-emoji.conf	\
+	70-no-bitmaps-except-emoji.conf	\
 	70-yes-bitmaps.conf		\
 	80-delicious.conf		\
 	90-synthetic.conf
diff --git a/conf.d/meson.build b/conf.d/meson.build
index 8e896118..b6d07ee0 100644
--- a/conf.d/meson.build
+++ b/conf.d/meson.build
@@ -36,6 +36,8 @@ conf_files = [
   '65-nonlatin.conf',
   '69-unifont.conf',
   '70-no-bitmaps.conf',
+  '70-no-bitmaps-and-emoji.conf',
+  '70-no-bitmaps-except-emoji.conf',
   '70-yes-bitmaps.conf',
   '80-delicious.conf',
   '90-synthetic.conf',
-- 
GitLab

